### PR TITLE
feat: Use CDK Diff to perform server mode integ tests

### PR DESCRIPTION
**Pull request by philasmar**
*Description of changes:*
* Add a StartDryRunDeployment operation in server mode that basically generates the CDK project and runs CDK Diff on that project
* In the integ test, we now check the generated CF template if it contains the correct settings.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs [hunk 1]
```diff
@@ -28,18 +33,17 @@ public class GetApplyOptionSettings : IDisposable
         private bool _isDisposed;
```
Comment (ashovlin) [inline test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs:33]: It won't let me comment on line 29 - but is `IntegrationTest` in the namespace still accurate?

If this is no longer relying on network calls and runs quickly (does it?), would `UnitTest` make it more likely that a developer working locally runs it?

File: test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs [hunk 2]
```diff
@@ -53,24 +57,20 @@ public GetApplyOptionSettings()
...
-            await Task.Delay(TimeSpan.FromSeconds(3));
-
-            await WaitUntilHelper.WaitUntil(async () =>
-            {
-                DeploymentStatus status = (await restApiClient.GetDeploymentStatusAsync(sessionId)).Status; ;
-                return status != DeploymentStatus.Executing;
-            }, TimeSpan.FromSeconds(1), TimeSpan.FromMinutes(15));
-
-            return (await restApiClient.GetDeploymentStatusAsync(sessionId)).Status;
+            var testCredentials = FallbackCredentialsFactory.GetCredentials();
```
Comment (ashovlin) [inline test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs:72]: Do we still need to access the user's default credentials if we're not actually deploying anything?

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs [hunk 3]
```diff
@@ -439,6 +439,39 @@ public async Task<IActionResult> GetCompatibility(string sessionId)
             return Ok(output);
         }
 
+        /// <summary>
+        /// Start a dry run of the deployment.
+        /// </summary>
+        [HttpPost("session/<sessionId>/dryrun")]
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:445]: I looked into that approach but the dry run should be an optional parameter. However, Swagger does not support the optional parameter and ends up generating the RestClient with a required parameter which breaks backwards compatibility and forces the IDE to always provide the dry run flag. That's why I went with the approach of adding a new operation.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs [hunk 4]
```diff
@@ -439,6 +439,39 @@ public async Task<IActionResult> GetCompatibility(string sessionId)
             return Ok(output);
         }
 
+        /// <summary>
+        /// Start a dry run of the deployment.
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:443]: A dry run of a deployment by our definition is generating the final CDK application along with the user properties, but not actually performing a "cdk deploy". I'll add that in the documentation.

File: test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs [hunk 5]
```diff
@@ -144,48 +145,31 @@ await restClient.ApplyConfigSettingsAsync(sessionId, new ApplyConfigSettingsInpu
...
 
-                Assert.Equal(_stackName, existingDeployment.Name);
-                Assert.Equal(appRunnerRecommendation.RecipeId, existingDeployment.RecipeId);
-                Assert.Equal(appRunnerRecommendation.Name, existingDeployment.RecipeName);
-                Assert.Equal(appRunnerRecommendation.ShortDescription, existingDeployment.ShortDescription);
-                Assert.Equal(appRunnerRecommendation.Description, existingDeployment.Description);
-                Assert.Equal(appRunnerRecommendation.TargetService, existingDeployment.TargetService);
-                Assert.Equal(DeploymentTypes.CloudFormationStack, existingDeployment.DeploymentType);
+                Assert.True(metadata.Settings.ContainsKey("VPCConnector"));
+                var vpcConnector = JsonConvert.DeserializeObject<VPCConnectorTypeHintResponse>(metadata.Settings["VPCConnector"].ToString());
```
[the remark this answers is not in the document]
Comment (philasmar) [reply test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs:164]: In simple terms yes, but the writing to JSON portion is done as part of the process of generating a CDK project and translating the Recommendation and setting configuration to a way the CDK project will understand. Then we are using CDK to compile the project and generate the CF template as it would during a normal deployment. Then we are inspecting the settings as part of the template. 
For the CDK assertion module you linked, I'm not quite sure how it would fit into this. Maybe we can have an offline discussion about it?

File: test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs (hunk 1 above)
Comment (philasmar) [reply test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs:33]: I'd still want to keep it under `IntergationTest` namespace as it is still technically an Integration Test even if it is not accessing the network. Developers could still run individual classes from the project without running the ones that require network calls. There are some tests in there that are similar.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs]: Done

File: src/AWS.Deploy.CLI/ServerMode/Models/StartDryRunDeploymentOutput.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+namespace AWS.Deploy.CLI.ServerMode.Models
+{
+    public class StartDryRunDeploymentOutput
+    {
+        public string CDKProject { get; set; }
+
+        public StartDryRunDeploymentOutput(string cdkProject)
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Models/StartDryRunDeploymentOutput.cs:18]: Done

**Review by 96malhar — APPROVED** [review]

File: test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply test/AWS.Deploy.CLI.IntegrationTests/ServerMode/GetApplyOptionSettings.cs:164]: Okay. Revisiting those docs, CDK's testing looks good for when you're in working in within the actual CDK project, but agree we're removed from that here.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:446]: Ah, and we don't have a request shape for `StartDeployment` so that may be breaking anyway - though if there was a time to do it, we're still in preview.

I'll defer to others who have been involved with the project longer, but still a little wary that we're expanding our public surface area this much to fix a testing issue.

Interested in @normj 's opinion if you get a chance.

[the remark this answers is not in the document]
Comment (normj) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:446]: I don't think you can have an optional resource path parameter. To make it optional you would need to make dryrun be a query string parameter. Can you try that and see if that works with the code generator?

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs [hunk 7]
```diff
@@ -439,6 +439,40 @@ public async Task<IActionResult> GetCompatibility(string sessionId)
...
 
+        /// <summary>
+        /// Start a dry run of the deployment, which performs all the steps in the process of performing a deployment except for running 'cdk bootstrap' and 'cdk deploy'.
+        /// This includes, but is not limited to, creating a deployment bundle and generating a CDK project.
+        /// </summary>
+        [HttpPost("session/<sessionId>/dryrun")]
+        [SwaggerOperation(OperationId = "StartDryRunDeployment")]
+        [SwaggerResponse(200, type: typeof(StartDryRunDeploymentOutput))]
+        [Authorize]
+        public async Task<IActionResult> StartDryRunDeployment(string sessionId)
```
Comment (normj) [inline src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:450]: This doesn't really seem like a dry run operation. It just returns back the path to the CDK project that is generated leaving to testing code to do anything meaningful. If we are going to alter our public API we should make it meaningful and something we can expand on.

This whole operation could to generate the CloudFormation template and return back the template to the user. That can be useful later in the CLI experience as well for users that just want the template and then have the template carried forward in their pipeline to later be deployed. Also the toolkit might want to have a button to view the template before deploying it.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (normj) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:446]: Actually based on my later comment I don't think we should try and squeeze this into the StartDeployment and instead have a specific method for returning back the CF template that would be deployed.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs [hunk 8]
```diff
@@ -439,6 +439,40 @@ public async Task<IActionResult> GetCompatibility(string sessionId)
...
+
+            var serviceProvider = CreateSessionServiceProvider(state);
+
+            var orchestratorSession = CreateOrchestratorSession(state);
+
+            var orchestrator = CreateOrchestrator(state, serviceProvider);
+
+            var cdkProjectHandler = CreateCdkProjectHandler(state, serviceProvider);
+
+            if (state.SelectedRecommendation == null)
```
Comment (normj) [inline src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:466]: There should be another error if the selected recommendation is not CDK based.

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs [hunk 9]
```diff
@@ -30,6 +32,14 @@ public async Task Execute()
             await _orchestrator.DeployRecommendation(_cloudApplication, _selectedRecommendation);
         }
 
+        public async Task<string> CreateCDKDeploymentProject(CdkProjectHandler cdkProjectHandler)
+        {
+            await CreateDeploymentBundle();
```
Comment (normj) [inline src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs:37]: We should just be forwarding into the `Orchestrator` which already has access to the `cdkprojecthandler` and we also don't have to add in `OrchestratorSession`.

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs [hunk 10]
```diff
@@ -30,6 +32,14 @@ public async Task Execute()
...
+            var cdkProject = await cdkProjectHandler.ConfigureCdkProject(_orchestratorSession, _cloudApplication, _selectedRecommendation);
+            await cdkProjectHandler.PerformCdkDiff(cdkProject);
```
Comment (normj) [inline src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs:39]: A method called `CreateCDKDeploymentProject` should also need to run `PerformCdkDiff`. I understand why you are doing it but we should be clear with our names and intents.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs (hunk 7 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs]: Done

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs (hunk 8 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:462]: done

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs (hunk 9 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs:40]: Done

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs (hunk 10 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs]: Done

File: src/AWS.Deploy.Orchestration/CdkProjectHandler.cs [hunk 11]
```diff
@@ -69,6 +73,21 @@ public async Task<string> ConfigureCdkProject(OrchestratorSession session, Cloud
             return cdkProjectPath;
         }
 
+        public async Task<string> PerformCdkDiff(string cdkProjectPath, CloudApplication cloudApplication)
```
Comment (ashovlin) [inline src/AWS.Deploy.Orchestration/CdkProjectHandler.cs:76]: Throughout this entire call stack - can you document what the return type is? Based on the function name assumed it'd be the `cdk diff` output, but looks like it's the full body of the CloudFormation template instead?

File: src/AWS.Deploy.CLI/AWS.Deploy.CLI.csproj [hunk 12]
```diff
@@ -43,6 +43,4 @@
     <None Include="..\..\icon.png" Pack="true" PackagePath="" />
   </ItemGroup>
 
-  <Import Project="..\AWS.Deploy.Constants\AWS.Deploy.Constants.projitems" Label="Shared" />
```
Comment (ashovlin) [inline src/AWS.Deploy.CLI/AWS.Deploy.CLI.csproj:46]: Why the change here and in the sln? I'm unsure how it's related.

File: src/AWS.Deploy.Orchestration/CdkProjectHandler.cs (hunk 11 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/CdkProjectHandler.cs:80]: I added documentation on the call stack

File: src/AWS.Deploy.CLI/AWS.Deploy.CLI.csproj (hunk 12 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/AWS.Deploy.CLI.csproj:46]: This is in response to Norm's comment about using the orchestrator to get to cdkprojecthandler. cdkprojecthandler is marked internal in the orchestrator, so i had to mark its internals visible to the CLI. By doing that I cause a conflict related to the constants project, so I had to remove it's reference from the CLI project as it can access them through the orchestrator.

**Review by ashovlin — APPROVED** [review]

File: src/AWS.Deploy.Orchestration/Properties/AssemblyInfo.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("AWS.Deploy.CLI")]
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/Properties/AssemblyInfo.cs:6]: CLI should not need `InternalsVisibleTo`

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs [hunk 14]
```diff
@@ -30,6 +30,30 @@ public async Task Execute()
...
+            if (_orchestrator._session == null)
+                throw new FailedToCreateCDKProjectException(DeployToolErrorCode.FailedToCreateCDKProject, $"We could not create a CDK deployment project due to a missing dependency '{nameof(_orchestrator._session)}'.");
+            if (_orchestrator._cdkProjectHandler == null)
+                throw new FailedToCreateCDKProjectException(DeployToolErrorCode.FailedToCreateCDKProject, $"We could not create a CDK deployment project due to a missing dependency '{nameof(_orchestrator._cdkProjectHandler)}'.");
+
+            await CreateDeploymentBundle();
+            var cdkProject = await _orchestrator._cdkProjectHandler.ConfigureCdkProject(_orchestrator._session, _cloudApplication, _selectedRecommendation);
+            try
+            {
+                return await _orchestrator._cdkProjectHandler.PerformCdkDiff(cdkProject, _cloudApplication);
```
Comment (normj) [inline src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs:49]: From the CLI project we should not be accessing internal members of the `Orchestrator`. I actually missed that we did in a previous PR made these fields internal. That was probably shouldn't have done.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Orchestration/Properties/AssemblyInfo.cs (hunk 13 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/Properties/AssemblyInfo.cs]: Fixed

File: src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs (hunk 14 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Tasks/DeployRecommendationTask.cs]: Fixed

**Review by normj — APPROVED** [review]
